### PR TITLE
APICompat - Use compatible assembly references

### LINKED ISSUE #35214: Investigate APICompat errors caused by assembly references re-used from the right-hand side during package validation
I.e. winforms when dropping net6.0 and net7.0 from their System.Drawing.Common package:

> D:\winforms\.dotnet\sdk\8.0.100-preview.7.23376.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API compatibility errors between 'lib/net6.0/System.Drawing.Common.dll' (d:\Nuget\system.drawing.common\7.0.0\system.drawing.common.7.0.0.nupkg) and 'lib/netstandard2.0/System.Drawing.Common.dll' (D:\winforms\artifacts\packages\Debug\Shipping\System.Drawing.Common.9.0.0-dev.nupkg):
1>D:\winforms\.dotnet\sdk\8.0.100-preview.7.23376.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP1002: Could not find matching assembly: 'System.Runtime.dll' in any of the search directories.
1>D:\winforms\.dotnet\sdk\8.0.100-preview.7.23376.3\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0008: Type 'System.Drawing.CharacterRange' does not implement interface 'System.IEquatable<System.Drawing.CharacterRange>' on lib/netstandard2.0/System.Drawing.Common.dll but it does on [Baseline] lib/net6.0/System.Drawing.Common.dll

### PR DESCRIPTION
**Pull request by ViktorHofer**
Fixes https://github.com/dotnet/sdk/issues/35214

When performing baseline validation, assembly references from the right hand side are considered when the left hand side doesn't provide references.

To make sure that only compatible (equal or newer) assembly references are considered, add a TFM check and find the nearest newer references.

Manually tested on dotnet/winforms and dotnet/runtime. This change should to be backported into .NET 8 as it reduces noise from https://github.com/dotnet/sdk/commit/c47af12ab3cd5f5ce35b221a55400bf6769bda9f which already is part of .NET 8.

### CODE REVIEW

**Comment by ViktorHofer** [comment]
Found an issue in runtime - investigating.

**Comment by ViktorHofer** [comment]
This change creates ~1500 errors in runtime for attributes and types that are internal in the current assembly but public in the baseline package assembly references (in the case of a net6.0 baseline assembly using the net8.0 assembly references from the current package):

> C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-dev\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error : API compatibility errors between 'lib/net6.0/System.Net.Http.Json.dll' (C:\Users\vihofer\.nuget\packages\system.net.http.json\8.0.0-rc.1.23415.6\system.net.http.json.8.0.0-rc.1.23415.6.nupkg) and 'lib/netstandard2.0/System.Net.Http
.Json.dll' (C:\git\runtime4\artifacts\packages\Debug\Shipping\System.Net.Http.Json.9.0.0-dev.nupkg): [C:\git\runtime4\src\libraries\System.Net.Http.Json\src\System.Net.Http.Json.csproj]


> C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-dev\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0014: Cannot remove attribute 'System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonType
Info or JsonSerializerContext, or make sure all of the required types are preserved.")' from 'System.Net.Http.Json.HttpClientJsonExtensions.PatchAsJsonAsync<TValue>(System.Net.Http.HttpClient, string?, TValue, System.Threading.CancellationToken)'. [C:\git\runtime4\src\libraries\System.Net.Http.Json\src\System.Net.Http.Json.csproj]
C:\Users\vihofer\.nuget\packages\microsoft.dotnet.apicompat.task\9.0.100-dev\build\Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0014: Cannot remove attribute 'System.Diagnostics.CodeAnalysis.RequiresDynamicCodeAttribute("JSON serialization and deserialization might require types that cannot be statically analyzed. Use the overload that takes a JsonTypeInfo
or JsonSerializerContext.")' from 'System.Net.Http.Json.HttpClientJsonExtensions.PatchAsJsonAsync<TValue>(System.Net.Http.HttpClient, string?, TValue, System.Threading.CancellationToken)'. [C:\git\runtime4\src\libraries\System.Net.Http.Json\src\System.Net.Http.Json.csproj]

`RequiresUnreferencedCodeAttribute` is internal in the current netstandard2.0 assembly but public in the baseline `net6.0` assembly that uses the `net8.0` assembly references.

**Comment by ViktorHofer** [comment]
@ericstj can you please take a look at the PR and the above comment? I'm not sure how we should treat these errors that arise from upgrading references from i.e. net6.0 to net8.0.

**Comment by ericstj** [comment]
Chatted about this offline.  These are all "real" errors.  They are showing up because we're comparing the netstandard2.0 assembly with the net6.0 assembly.  We'd see the same errors if he had the "real" references for `net6.0`.

We're now able to see that the net6.0 assembly applied attributes that are not equivalent to the netstandard2.0 attributes.  In the netstandard2.0 case those were internal attributes, so they aren't part of the comparison.  I don't think we want try to hide those - since if someone did this in a "real" case it would be a problem.  In our case we aren't actually trying to claim that the `netstandard2.0` implementation is actually compatible with the `net6.0` assembly - so we should try to make it as easy as possible to communicate this intent and not introduce hacks/complexity to deal with the noise.

**Comment by ericstj** [comment]
Let me know when this is ready to go and I'll have another look.  Any possibility to include tests?

**Comment by ViktorHofer** [comment]
@ericstj PTAL. Here's the corresponding runtime change: https://github.com/dotnet/runtime/pull/93129

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs [hunk 1]
```diff
@@ -78,17 +78,22 @@ public bool IsErrorSuppressed(Suppression error)
                 // - DiagnosticId, Target, IsBaselineSuppression
                 Suppression globalTargetSuppression = new(error.DiagnosticId, error.Target, isBaselineSuppression: error.IsBaselineSuppression);
 
+                // - Left, Right, IsBaselineSuppression
+                Suppression globalLeftRightSuppression = new(string.Empty, left: error.Left, right: error.Right, isBaselineSuppression: error.IsBaselineSuppression);
+
                 // - DiagnosticId, Left, Right, IsBaselineSuppression
-                Suppression globalLeftRightSuppression = new(error.DiagnosticId, left: error.Left, right: error.Right, isBaselineSuppression: error.IsBaselineSuppression);
+                Suppression globalDiagnosticIdLeftRightSuppression = new(error.DiagnosticId, left: error.Left, right: error.Right, isBaselineSuppression: error.IsBaselineSuppression);
```
[the remark this answers is not in the document]
Comment (ericstj) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:85]: We should not manually loop - that's trading off significant CPU for small amount of memory.

A consideration for the future might be making Suppression an (immutable) struct.  I think that would address allocation concerns.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs [hunk 2]
```diff
@@ -57,9 +57,17 @@ private Suppression()
         /// the property won't be serialized to keep the baseline file minimal. The method's name is important as
         /// XmlSerializer will look for methods called ShouldSerializeX to determine if properties should be serialized.
         /// </summary>
-        /// <returns>Returns true if IsBaselineSuppression should be serialized</returns>
+        /// <returns>Returns true when IsBaselineSuppression should be serialized.</returns>
         public bool ShouldSerializeIsBaselineSuppression() => IsBaselineSuppression;
 
+        /// <summary>
+        /// Only serialize the DiagnosticId property when it is not empty. This can happen when a global suppression
```
Comment (ericstj) [inline src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs:64]: Does this mean that a suppression file with an empty diagnostic ID will lose that suppression when regenerating it?

**Review by ericstj — COMMENTED** [review]
I wish we could test the reference cases, but I understand that the current test methods make this hard and you've mitigated that gap with manual test of dotnet/runtime.

I just had one question about the diagnostic behavior.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs:64]: No. It means that a global suppression without a diagnostic id that is already stored in the suppression file does get a `<Diagnostic />` entry added when re-generating the suppression file.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ViktorHofer) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:85]: I need to benchmark this but I wouldn't be surprised if the current code with the 3 object instantiations and the hashset contains calls is slower than a list foreach without an object instantiation for under a thousand entries.

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs (hunk 2 above)
Comment (ericstj) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/Suppression.cs:64]: I see.  Got it, this keeps the generated file concise.  Makes sense.  Thanks for the explanation.

**Review by ericstj — APPROVED** [review]

File: src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericstj) [reply src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Logging/SuppressionEngine.cs:85]: Might be an interesting measurement.  Where I was coming from with my comment is that hashset with O(1) lookup would be much faster than list with O(n), especially with a size of n like 1000.  Also consider a common use case is to suppress everything which means we do this n times where our matches will be evenly distributed.  Might depend on the similarity of the items and how much work the `Equals` method needs to do for each item compared to the cost of `GetHashCode` on look up item.  Object creation isn't all that expensive and changing to struct would eliminate work for the GC - these objects are just simple types containing references to existing strings so their constructors do next to no work at all.

**Pull request merged**